CRM/Automateddirectdebit/Job/DirectDebitEvents/PaymentCollectionEvent.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+/**
+ * Responsible for emitting direct debit payment collection event (hook).
+ * The dispatched event/hook is called: `automateddirectdebit_PaymentCollectionEvent`
+ * where 3rd party payment processor extensions (such as GoCardless extension) can implement it
+ * to create and collect payments.
+ *
+ */
+class CRM_Automateddirectdebit_Job_DirectDebitEvents_PaymentCollectionEvent {
+
+  public function run() {
+    $pendingInvoicesQuery = $this->buildPendingInvoicesQuery();
+    $pendingInvoices = CRM_Core_DAO::executeQuery($pendingInvoicesQuery->toSQL());
+
+    while ($pendingInvoices->fetch()) {
+      $pendingInvoiceData = $pendingInvoices->toArray();
+      $invoiceTotalPaidAmount = CRM_Core_BAO_FinancialTrxn::getTotalPayments($pendingInvoiceData['contribution_id'], TRUE);
+      if ($invoiceTotalPaidAmount >= $pendingInvoiceData['total_amount']) {
+        continue;
+      }
+
+      $pendingInvoiceData['charge_amount'] = $pendingInvoiceData['total_amount'] - $invoiceTotalPaidAmount;
+
+      $this->markContributionExternalPaymentToBeInProgress($pendingInvoiceData['contribution_id']);
+      $this->dispatchPaymentCollectionEventHook($pendingInvoiceData);
+    }
+  }
+
+  /**
+   * Builds the query to fetch the contributions (invoices)
+   * that match the criteria of invoices that we
+   *
+   * @return CRM_Utils_SQL_Select
+   */
+  public function buildPendingInvoicesQuery() {
+    $recurContributionStatusesToProcess = implode(',', $this->getRecurContributionStatusesIdsToProcess());
+    $mandateActiveStatusId = 1;
+    $failureRetryCount = \Civi::settings()->get('automateddirectdebit_paymentplan_payment_collection_retry_count');
+    $todaysDate = date('Y-m-d 00:00:00');
+
+    $query = CRM_Utils_SQL_Select::from('civicrm_contribution c')
+      ->join('cr', 'INNER JOIN civicrm_contribution_recur cr ON c.contribution_recur_id = cr.id')
+      ->join('mandate', 'INNER JOIN civicrm_value_external_dd_mandate_information mandate ON cr.id = mandate.entity_id')
+      ->join('ppea', 'INNER JOIN civicrm_value_payment_plan_extra_attributes ppea ON cr.id = ppea.entity_id')
+      ->join('epi', 'LEFT JOIN civicrm_value_external_dd_payment_information epi ON c.id = epi.entity_id')
+      ->where("mandate.mandate_id IS NOT NULL")
+      ->where("mandate.mandate_status = {$mandateActiveStatusId}")
+      ->where('ppea.is_active = 1')
+      ->where("cr.contribution_status_id IN ({$recurContributionStatusesToProcess})")
+      ->where('mandate.next_available_payment_date IS NOT NULL')
+      ->where('c.receive_date >= mandate.next_available_payment_date')
+      ->where("c.receive_date <= '{$todaysDate}'")
```

Comment:
I think the received_date might have time. For example:

received date: `2023-07-24 18:23:18`
today's date: `2023-07-24 00:00:00`

So, I am not sure if the logic would work because when comparing with the time, the received date could always appear greater than today's date, even though both received date and today's date are on the same day.

See the example:
`2023-07-24 18:23:18` <= `2023-07-24 00:00:00`

Perhaps, we should omit the time when comparing the two dates?

Reply:
you are right, I wanted to use DATE() function on the recevie_date but that would end up making the index on it useless, but I think I have a solution, I will use this instead:

` ->where("c.receive_date < DATE_ADD(CURDATE(), INTERVAL 1 DAY)")`